File: vol_007/p785.cpp
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <assert.h>
#include <iostream>
#include <string>
#include <vector>
#include <stack>
#include <set>

bool is_end_line(const std::string& s)
{
    for (char c : s) if (c != '_') return false;
    return !s.empty();
}

int main(int argc, char **argv)
{
    for (;;)
    {
        std::vector<std::string> grid;
        std::string end_line, line;
        for (;;)
        {
            if (!std::getline(std::cin,line)) goto end; // EOF
            if (is_end_line(line))
            {
                end_line = line;
                break;
            }
            grid.push_back(line);
        }
        std::vector<std::vector<bool> > marked;
        for (auto r : grid) // generate
            marked.push_back(std::vector<bool>(r.length(),false));
        // find contour char
        char contour = '\0';
        for (auto r : grid)
            for (auto c : r)
                if (c != '_' and c != ' ')
                {
                    contour = c;
                    goto done1;
                }
        done1:
        assert(contour != '\0'); // contour was found
        // region filling loop
        for (size_t r = 0; r != marked.size(); ++r)
            for (size_t c = 0; c != marked[r].size(); ++c)
            {
                if (marked[r][c]) continue; // was part of previous region
                if (grid[r][c] == contour) continue; // not inside region
                std::stack<std::pair<size_t,size_t> > stack;
                stack.push(std::make_pair(r,c)); // for iterative flood fill
                char marking_char = '\0';
                std::set<std::pair<size_t,size_t> > region; // points inside
                while (!stack.empty())
                {
                    std::pair<size_t,size_t> loc = stack.top();
                    stack.pop();
                    if (loc.first >= marked.size() // bounds check
                        or loc.second >= marked[loc.first].size()) continue;
                    if (grid[loc.first][loc.second] == contour) continue;
                    if (marked[loc.first][loc.second]) continue;
                    if (grid[loc.first][loc.second] != ' ') // marking char
                    {
                        if (marking_char == '\0') // first instance
                            marking_char = grid[loc.first][loc.second];
                        else assert(grid[loc.first][loc.second]
                                    == marking_char); // should be same
                    }
                    marked[loc.first][loc.second] = true;
                    region.insert(loc);
                    // DFS to adjacent cells
                    stack.push(std::make_pair(loc.first+1,loc.second));
                    stack.push(std::make_pair(loc.first,loc.second+1));
                    stack.push(std::make_pair(loc.first-1,loc.second));
                    stack.push(std::make_pair(loc.first,loc.second-1));
                }
                if (marking_char != '\0') // fill region
                    for (auto loc : region)
                        grid[loc.first][loc.second] = marking_char;
            }
        // print results
        for (auto r : grid) printf("%s\n",r.c_str());
        printf("%s\n",end_line.c_str());
    }
    end:
    return 0;
}

File: vol_007/p785.py
import sys

while True:
    grid = [] # 2d array of characters
    end_line = ''
    reach_eof = False
    while True: # read input lines
        try: line = input()
        except EOFError:
            reach_eof = True
            break
        if set(line) == set('_'):
            end_line = line
            break
        grid.append(list(line))
    if reach_eof: break
    marked = [] # mark visited locations in flood fill
    for r in grid:
        marked.append([False]*len(r))
    contour = '' # find edge character used for contours
    for r in grid:
        found = False
        for c in r:
            if c != '_' and c != ' ':
                contour = c
                found = True
                break
        if found: break
    assert contour != ''
    for r in range(len(marked)): # perform region filling
        for c in range(len(marked[r])):
            if marked[r][c]: continue # already part of a filled region
            if grid[r][c] == contour: continue # not inside a region
            stack = [(r,c)] # grid positions to search, iterative DFS
            marking_char = '' # the non space char in this region
            region = set() # positions in this region
            while len(stack): # this should not mark the contours
                rr, cc = stack.pop()
                if rr < 0 or rr >= len(marked) \
                    or cc < 0 or cc >= len(marked[rr]): # out of bounds
                    continue
                if grid[rr][cc] == contour: continue # edge of region
                if marked[rr][cc]: continue # was already marked
                if grid[rr][cc] != ' ': # found marking char
                    if marking_char == '': marking_char = grid[rr][cc]
                    else: assert grid[rr][cc] == marking_char
                marked[rr][cc] = True # mark visited cell
                region.add((rr,cc)) # position in current region
                # DFS to adjacent cells
                stack += [(rr+1,cc),(rr,cc+1),(rr-1,cc),(rr,cc-1)]
            if marking_char == '': continue # unmarked region
            else: # fill region
                for r,c in region:
                    grid[r][c] = marking_char
    # print results
    for r in grid:
        print(''.join(r))
    print(end_line)

